Add PSA for blocking `.host` on pod probes

tssurya opened the pull request:
#### What type of PR is this?

/kind bug
/kind feature

#### What this PR does / why we need it:

This PR aims to add the following fields as part of PSA so that users can be blocked/warned when they create probes with the `.Host` field set.

```
* spec.containers[*].livenessProbe.httpGet.host
* spec.containers[*].readinessProbe.httpGet.host
* spec.containers[*].startupProbe.httpGet.host
* spec.containers[*].livenessProbe.tcpSocket.host
* spec.containers[*].readinessProbe.tcpSocket.host
* spec.containers[*].startupProbe.tcpSocket.host
* spec.containers[*].lifecycle.postStart.tcpSocket.host // Deprecated. TCPSocket is NOT supported as a LifecycleHandler and kept for backward compatibility.
* spec.containers[*].lifecycle.preStop.tcpSocket.host // Deprecated. TCPSocket is NOT supported as a LifecycleHandler and kept for backward compatibility.
* spec.containers[*].lifecycle.postStart.httpGet.host
* spec.containers[*].lifecycle.preStop.httpGet.host
* spec.initContainers[*].livenessProbe.httpGet.host
* spec.initContainers[*].readinessProbe.httpGet.host
* spec.initContainers[*].startupProbe.httpGet.host
* spec.initContainers[*].livenessProbe.tcpSocket.host
* spec.initContainers[*].readinessProbe.tcpSocket.host
* spec.initContainers[*].startupProbe.tcpSocket.host
* spec.initContainers[*].lifecycle.postStart.tcpSocket.host // Deprecated. TCPSocket is NOT supported as a LifecycleHandler and kept for backward compatibility.
* spec.initContainers[*].lifecycle.preStop.tcpSocket.host // Deprecated. TCPSocket is NOT supported as a LifecycleHandler and kept for backward compatibility.
* spec.initContainers[*].lifecycle.postStart.httpGet.host
* spec.initContainers[*].lifecycle.preStop.httpGet.host
```

#### Which issue(s) this PR fixes:

Fixes #99425
Fixes https://github.com/kubernetes/enhancements/issues/4940

#### Special notes for your reviewer:

This is my first PR to pod-security-admission area, so I might need some hand-holding, but I will mark the PR as draft till I get all the pieces done.

So one question I have is should I fix the tests on sig-node side like I have done here today by making them use the privileged mode? Or should I do something like https://github.com/kubernetes/kubernetes/pull/132916/commits/77fa562ae7ee3c6aec33ab1d19a2e259d1058b40

Ran the following tests locally:

- UPDATE_POD_SECURITY_FIXTURE_DATA=true go test -v ./test/... -run TestFixtures
- go test -v ./test/... -run TestFixtures
- go test -v ./policy/... -run TestHostProbesAndHostLifecycle
- go test -v ./policy/... -timeout 30s
- go test -v ./test/... -timeout 30s
- make test-integration WHAT=./test/integration/auth GOFLAGS="-v" KUBE_TEST_ARGS='-run ^TestPodSecurity$$'

#### Does this PR introduce a user-facing change?


```release-note
The `baseline` and `restricted` pod security admission levels now block setting the `host` field on probe and lifecycle handlers
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:



Please use the following format for linking documentation:
- KEP: https://github.com/kubernetes/enhancements/pull/4558 and https://github.com/kubernetes/enhancements/issues/4940
- Usage: <link>
- Other doc: https://github.com/kubernetes/website/pull/51498

```docs

```

tssurya committed:
Add PSA for blocking .host field on pod probes and lifecycle

Signed-off-by: Surya Seetharaman <suryaseetharaman.9@gmail.com>

deads2k commented on staging/src/k8s.io/pod-security-admission/policy/check_probes.go:
```diff
@@ -0,0 +1,104 @@
+		Level: api.LevelBaseline,
+		Versions: []VersionedCheck{
+			{
+				MinimumVersion: api.MajorMinorVersion(1, 0),
```
this version is the version where the restriction is enforced.

tssurya replied:
yup updated this to 1.34

ibihim commented:
/triage accepted

tssurya commented:
/remove-lifecycle stale

tssurya commented:
> Hello @tssurya @liggitt This PR has not been updated for 4 months, so I'd like to check what's the status. If there's anything we can do, please let us know. The code freeze is starting 02:00 UTC Friday 25th July 2025 (about 4 weeks from now). Please make sure the PR has both lgtm and approved labels before the code freeze. Thanks!

hi! thanks for the ping, yes its on my radar and it will be ready this week for reviews. I'll make sure its done before the deadline. I am working on getting this ready this week

tssurya commented on staging/src/k8s.io/pod-security-admission/policy/check_hostProbesAndhostLifecycle_test.go:
```diff
@@ -0,0 +1,235 @@
+						Name: "a",
+						LivenessProbe: &corev1.Probe{
+							ProbeHandler: corev1.ProbeHandler{
+								HTTPGet: &corev1.HTTPGetAction{Host: "invalid.host"},
```
@danwinship : so host field can be hostnames or ips right?

tssurya commented:
@danwinship @deads2k this PR is ready for reviews. The spot for unit tests was easy to find.
But integration tests.. I am a bit stuck on where to start a PR for those.

tssurya committed:
Add new fixture testData for 1.34 PSA

This commit adds the fixture tests for the
new .host field restrictions on probe
and lifecycle handlers.

ran UPDATE_POD_SECURITY_FIXTURE_DATA=true go test -v ./test/... -run TestFixtures

Signed-off-by: Surya Seetharaman <suryaseetharaman.9@gmail.com>

tssurya committed:
update the minimal version to fix unit/fixture tests

Signed-off-by: Surya Seetharaman <suryaseetharaman.9@gmail.com>

tssurya commented:
` Container Lifecycle Hook when create a pod with lifecycle hook should execute poststart exec hook properly` looks like I need to figure out if I should change this to localhost or do something else.. i thought the default was always podIP if host wasn't set.. so wonder why these are failing. need to look into that

tssurya committed:
Copy test fixtures to 1.34 from 1.32

Signed-off-by: Surya Seetharaman <suryaseetharaman.9@gmail.com>

tssurya committed:
Copy test fixtures to 1.33 from 1.32

Signed-off-by: Surya Seetharaman <suryaseetharaman.9@gmail.com>

tssurya commented:
@danwinship : managed to get unit tests to pass! did the trick with copying over the folders... :)

tssurya commented:
hmm the sig-node tests are still failing.. looks like these tests are trying to set the host field to another server pod's IP which isn't supported... let me see if I can modify these tests to point towards the same client podIP itself? or should I totally remove these tests I wonder...

tssurya committed:
Tests using .host field in probes must be at priviledged level

The sig-node tests have scenarios of doing probes and
lifecycle handler tests with post-start and pre-stop hooks
setting the host field to be another pod.

In baseline level such things won't be allowed because of
the PSA rules we are adding in this PR. So unsetting
the host field means it uses the podIP of self for doing
the checks and using that in the pre-stop and post-start
hooks is tricky because of the timing issues with when the
container is actually up v/s running the test.

So I have changed the tests to be privileded for them to
use the .host fields if they desire to.

See https://github.com/kubernetes/kubernetes/issues/133091
which is an issue opened to properly refactor these tests.

Signed-off-by: Surya Seetharaman <suryaseetharaman.9@gmail.com>

tssurya commented on test/e2e/common/node/lifecycle_hook.go:
```diff
@@ -41,7 +41,7 @@ import (
 var _ = SIGDescribe("Container Lifecycle Hook", func() {
 	f := framework.NewDefaultFramework("container-lifecycle-hook")
-	f.NamespacePodSecurityLevel = admissionapi.LevelBaseline
+	f.NamespacePodSecurityLevel = admissionapi.LevelPrivileged
```
exactly, I am unable to fix the race between the process inside the container being ready v/s lifecycle hooks succeeding when we don't set hots which means the self pod's status.IPs is picked and doing a lifecycle check to itself isn't really a thing?

@mrunalp @haircommander what is the right way to fix this any ideas?

tssurya replied:
I'll open an issue tomorrow and link it here as a comment.

liggitt replied:
hmm... I guess it's not blocking, but please add a comment here about *why* this test is getting elevated permissions, and link to the issue / comment https://github.com/kubernetes/kubernetes/issues/133091#issuecomment-3097137762 for an explanation and a todo to resolve

tssurya replied:
yup will do - I'll add a comment to the code snippet and mention the details

tssurya replied:
I've added a FIXME comment pointing to the issue

tssurya commented on staging/src/k8s.io/pod-security-admission/policy/check_hostProbesAndhostLifecycle.go:
```diff
@@ -0,0 +1,155 @@
+)
+
+/*
+Host field must be forbidden in the probes and lifecycle handlers.
```
done

tssurya commented:
/test pull-kubernetes-e2e-kind

liggitt commented on staging/src/k8s.io/pod-security-admission/policy/check_hostProbesAndhostLifecycle.go:
```diff
@@ -0,0 +1,159 @@
+	}
+}
+
+func hostProbesAndHostLifecycleV1Dot34(podMetadata *metav1.ObjectMeta, podSpec *corev1.PodSpec) CheckResult {
+	var badContainers []string
+	forbidden := sets.New[string]()
+	allowed := sets.New[string](allowedLocalHostIPv4, allowedLocalHostIPv6)
+	visitContainers(podSpec, func(container *corev1.Container) {
```
I can't remember, do we also need to block use of HTTPGet.HTTPHeader[*].Name = case-insensitive("host")?

tssurya replied:
> I think this can be a lot simpler / more concise / more performant where the normal case will return nil for successful non-violating cases. Something like this:
> 
> ```go
> var allowedHostValues = sets.New[string]("", "127.0.0.1", "::1", "localhost")
> 
> func hostProbesAndHostLifecycleV1Dot34(podMetadata *metav1.ObjectMeta, podSpec *corev1.PodSpec) CheckResult {
>   badContainers := sets.New[string]()
>   badHosts := sets.New[string]()
>   visitContainers(podSpec, func(container *corev1.Container) {
>     if badHosts := getForbiddenHosts(container.LivenessProbe); len(badHosts) > 0 {
>       badContainers.Insert(container.Name)
>       badHosts.Insert(badHosts...)
>     }
>     if badHosts := getForbiddenHosts(container.ReadinessProbe); len(badHosts) > 0 {
>       badContainers.Insert(container.Name)
>       badHosts.Insert(badHosts...)
>     }
>     if badHosts := getForbiddenHosts(container.StartupProbe); len(badHosts) > 0 {
>       badContainers.Insert(container.Name)
>       badHosts.Insert(badHosts...)
>     }
>   })
> 
>   if len(badContainers) > 0 {
> 		return CheckResult{
> 			Allowed:         false,
> 			ForbiddenReason: "forbidden probe hosts",
> 			ForbiddenDetail: fmt.Sprintf(
> 				"%s %s %s %s %s",
> 				pluralize("container", "containers", len(badContainers)),
> 				joinQuote(sets.List(badContainers)),
> 				pluralize("uses", "use", len(badContainers)),
> 				pluralize("invalid probe host", "invalid probe hosts", len(badHosts)),
> 				joinQuote(sets.List(badHosts)),
> 			),
> 		}
>   }
>   return CheckResult{Allowed: true}
> }
> 
> func getForbiddenHosts(probe *corev1.Probe) []string {
>   var badHosts []string
>   if probe == nil {
>     return nil
>   }
>   if probe.HTTPGet != nil && !allowedHostValues.Has(probe.HTTPGet.Host) {
>     badHosts = append(badHosts, probe.HTTPGet.Host)
>   }
>   if probe.TCPSocketAction != nil && !allowedHostValues.Has(probe.TCPSocketAction.Host) {
>     badHosts = append(badHosts, probe.TCPSocketAction.Host)
>   }
>   return badHosts
> }
> ```

I actually had this in my first version and then in my second version I changed it to also include lifecycle hooks.. so passing just the probes field means I would need another util to pass the lifecycle hook constructs but yes I can change this code according to the suggestion made here.

liggitt replied:
Ah, I missed the lifecycle bit... yeah, go ahead and do that. Flatter / smaller / simpler methods are easier to reason about.

tssurya commented on staging/src/k8s.io/pod-security-admission/policy/check_hostProbesAndhostLifecycle.go:
```diff
@@ -0,0 +1,159 @@
+*/
+
+const (
+	allowedLocalHostIPv4 = "127.0.0.1"
```
hmm good catch I think we should include `localhost` as well

liggitt replied:
looking at the kubelet probe code, doesn't 127.0.0.1 mean the kubelet is making requests to the host network, not the pod IP? why is 127.0.0.1 allowed for pods which aren't supposed to have host network access?

tssurya replied:
correct the 127.0.0.1 is localhost of the network and not the podIP.
Reasoning behind allowing this specific probe was because there is a use case where people are using localhost probes where public IP of the node might be blocked by firewall but apiserver being host-networked is created with such healthcheck probes in our tests: https://github.com/kubernetes/kops/blob/5dd2f468b46fda43f3a63ba1e6dc7c55c21919eb/nodeup/pkg/model/kube_apiserver.go#L603 

https://github.com/kubernetes/enhancements/pull/4942#discussion_r2123759432 was the discussion which lead to https://github.com/kubernetes/enhancements/pull/5444

liggitt replied:
I left a comment at https://github.com/kubernetes/enhancements/pull/5444/files#r2225633269, but I don't think that scenario makes sense.

The linked example of the kube-apiserver pod used `hostNetwork: true`, and that pod would be rejected by the baseline policy. hostNetwork and hostPort pods are not allowed in the baseline policy.

https://github.com/kubernetes/kubernetes/blob/9148694d4bcaf620a5190899f040213f8f44561c/staging/src/k8s.io/pod-security-admission/policy/check_hostNamespaces.go#L32

https://github.com/kubernetes/kubernetes/blob/9148694d4bcaf620a5190899f040213f8f44561c/staging/src/k8s.io/pod-security-admission/policy/check_hostPorts.go#L35-L36

tssurya replied:
oh! I wasn't aware `hostNetwork:true` was totally blocked in baseline... hmm then I'm not sure if we need to carry these exceptions around for localhost.. I'm not very familiar with "kops" is that a specific SIG or do those tests run somewhere in CI to ensure I don't break them? I'm sure I didn't hit those in CI

@danwinship thoughts?

danwinship replied:
Ben mentioned "kops" as an example but then we noticed that the deployment stuff in k/k/cluster/gce uses probes with `host: 127.0.0.1` too. (We just hadn't noticed originally because I was grepping for YAML but it uses JSON.) Anyway, you can see in the `pull-kubernetes-e2e-gce` logs that it _is_ using that configuration:

```text
Jul 23 11:01:55.391392 bootstrap-e2e-master kubelet[9706]: I0723 11:01:55.391283    9706 http.go:116] Probe succeeded for https://127.0.0.1:443/readyz, Response: {200 OK 200 HTTP/2.0 2 0 map[Audit-Id:[543b32c6-a277-4ab1-8a7b-6d4c6deecf44] Cache-Control:[no-cache, private] Content-Length:[2] Content-Type:[text/plain; charset=utf-8] Date:[Wed, 23 Jul 2025 11:01:55 GMT] X-Content-Type-Options:[nosniff] X-Kubernetes-Pf-Flowschema-Uid:[df4ea8eb-1728-43f3-bc92-a5a7d5f80319] X-Kubernetes-Pf-Prioritylevel-Uid:[46858e6e-6650-4df6-92fc-4887e2df4dfe]] 0xc0008a43e0 2 [] false false map[] 0xc0004423c0 0xc0015dc240}
Jul 23 11:01:55.392031 bootstrap-e2e-master kubelet[9706]: I0723 11:01:55.391990    9706 prober.go:111] "Probe succeeded" probeType="Readiness" pod="kube-system/kube-apiserver-bootstrap-e2e-master" podUID="e3d9a1ab458c329cf1ba12d4d6721aff" containerName="kube-apiserver"
```

so, if e2e passes here with all `host`-in-probes blocked, then we're not breaking that use case.

tssurya replied:
In my latest push I removed support for allowing localhost values. I think we are looking good in CI e2e's. I'll clean up the PR and do a final push and also close the enhancement update PR

adilGhaffarDev commented:
Hello @tssurya @liggitt 
Appreciate all of your efforts with this PR! Is the plan still to resolve it for v1.34 release?
If so, a gentle reminder that the code freeze has started [02:00 UTC Friday 25th July 2025](https://everytimezone.com/s/a2c01c54) . Please make sure any PRs have both lgtm and approved labels ASAP, and file an [Exception](https://github.com/kubernetes/sig-release/blob/master/releases/EXCEPTIONS.md) if you haven't done it yet.
Thanks!

tssurya commented:
@adilGhaffarDev : I am working towards the 25th deadline... almost everything here is straightforward except https://github.com/kubernetes/kubernetes/pull/125271#discussion_r2207807261 I'd like to hear @liggitt 's thoughts on if he is OK with a FUP via the issue or if the tests refactor need to be part of this PR.. rest of the comments are straightforward to address

tssurya commented on staging/src/k8s.io/pod-security-admission/policy/check_hostProbesAndhostLifecycle.go:
```diff
@@ -0,0 +1,159 @@
+	return CheckResult{Allowed: true}
+}
+
+func getForbiddenHosts(host string, allowed sets.Set[string]) sets.Set[string] {
```
changed the structure to match your earlier suggestion..

tssurya commented:
The first push was only plain rebase on master.
The second push contains the fix: https://github.com/kubernetes/kubernetes/compare/470626a6c9326dc7269cf90e11729928a59a3a38..bd8c7d6fc321c9c69ed1877a503b65098c492d08

tssurya commented:
https://prow.k8s.io/view/gs/kubernetes-ci-logs/pr-logs/pull/125271/pull-kubernetes-e2e-kind/1948068821893910528

```
Kubernetes e2e suite: [It] [sig-cli] Kubectl client Kubectl events should show event when pod is created expand_less	6s
{ failed [FAILED] unexpected non-timeout WARNING event fired, got: LAST SEEN   TYPE      REASON   OBJECT                   MESSAGE
3s          Warning   Failed   Pod/e2e-test-httpd-pod   Failed to pull image "registry.k8s.io/e2e-test-images/httpd:2.4.38-4": pull QPS exceeded
3s          Warning   Failed   Pod/e2e-test-httpd-pod   Error: ErrImagePull
 
In [It] at: k8s.io/kubernetes/test/e2e/kubectl/kubectl.go:2052 @ 07/23/25 17:19:56.482
}
```
 looks unrelated?

liggitt reviewed (commented):
a couple nits, then lgtm

please also open the documentation PR to update https://kubernetes.io/docs/concepts/security/pod-security-standards/#baseline for the dev-1.34 branch in https://github.com/kubernetes/website/

tssurya commented on staging/src/k8s.io/pod-security-admission/policy/check_hostProbesAndhostLifecycle.go:
```diff
@@ -0,0 +1,148 @@
+* spec.initContainers[*].lifecycle.preStop.tcpSocket.host // Deprecated. TCPSocket is NOT supported as a LifecycleHandler and kept for backward compatibility.
+* spec.initContainers[*].lifecycle.postStart.httpGet.host
+* spec.initContainers[*].lifecycle.preStop.httpGet.host
+*/
```
done

liggitt commented:
/lgtm
/approve

tssurya commented:
@michaelasp : if the tests are using '.Host' in the life cycle hook then yes I might have broken them :(

But how come I didn't catch these in PR's CI?

danwinship commented:
It seems like this might be the first new PSA rule added since [KEP-4330: Compatibility Versions in Kubernetes](https://github.com/kubernetes/enhancements/blob/master/keps/sig-architecture/4330-compatibility-versions/README.md)?

I _think_ the answer here is that PSA application needs to be based on the `--emulated-version` rather than the actual apiserver version. @liggitt?

AkihiroSuda commented:
This one is still failing since 2025-07-24:
- https://github.com/kubernetes/kubernetes/issues/133330

```
• [FAILED] [6.590 seconds]
[sig-node] Container Lifecycle Hook when create a pod with lifecycle hook [It] should execute poststart https hook properly [MinimumKubeletVersion:1.23] [NodeConformance] [sig-node, NodeConformance]
k8s.io/kubernetes/test/e2e/common/node/lifecycle_hook.go:194
   Timeline >>
  STEP: Creating a kubernetes client @ 07/24/25 07:49:56.477
  I0724 07:49:56.477650 2815 util.go:453] >>> kubeConfig: /home/prow/go/src/k8s.io/kubernetes/_rundir/188a9770-7f1e-4d01-8648-e5dce8a2ca03/kubeconfig
  STEP: Building a namespace api object, basename container-lifecycle-hook @ 07/24/25 07:49:56.478
  STEP: Waiting for a default service account to be provisioned in namespace @ 07/24/25 07:49:56.561
  STEP: Waiting for kube-root-ca.crt to be provisioned in namespace @ 07/24/25 07:49:56.569
  STEP: create the container to handle the HTTPGet hook request. @ 07/24/25 07:49:56.59
  STEP: create the pod with lifecycle hook @ 07/24/25 07:50:02.706
  I0724 07:50:02.712557 2815 lifecycle_hook.go:101] Unexpected error: Error creating Pod: 
      <*errors.StatusError | 0xc00063c000>: 
      pods "pod-with-poststart-https-hook" is forbidden: violates PodSecurity "baseline:latest": probe or lifecycle host (container "pod-with-poststart-https-hook" uses probe or lifecycle host "10.244.0.7")
      {
          ErrStatus: 
              code: 403
              details:
                kind: pods
                name: pod-with-poststart-https-hook
              message: 'pods "pod-with-poststart-https-hook" is forbidden: violates PodSecurity
                "baseline:latest": probe or lifecycle host (container "pod-with-poststart-https-hook"
                uses probe or lifecycle host "10.244.0.7")'
              metadata: {}
              reason: Forbidden
              status: Failure,
      }
```
